numpy/core/shape_base.py

```diff
@@ -365,78 +365,83 @@ def stack(arrays, axis=0, out=None):
-                    for xi in x
-                ), **kwargs)
-        return f(x, **kwargs)
-
-    def walk(self, x, index=()):
-        """
-        Iterate over x, yielding (index, value, entering), where
-
-        * ``index``: a tuple of indices up to this point
-        * ``value``: equal to ``x[index[0]][...][index[-1]]``. On the first iteration, is
-                     ``x`` itself
-        * ``entering``: bool. The result of ``recurse_if(value)``
-        """
-        do_recurse = self.recurse_if(x)
-        yield index, x, do_recurse
-
-        if not do_recurse:
-            return
-        for i, xi in enumerate(x):
-            # yield from ...
-            for v in self.walk(xi, index + (i,)):
-                yield v
+    def format_index(index):
+        idx_str = ''.join('[{}]'.format(i) for i in index if i is not None)
+        return 'arrays' + idx_str
+    if type(arrays) is tuple:
+        # not strictly necessary, but saves us from:
+        #  - more than one way to do things - no point treating tuples like
+        #    lists
+        #  - horribly confusing behaviour that results when tuples are
+        #    treated like ndarray
+        raise TypeError(
+            '{} is a tuple. '
+            'Only lists can be used to arrange blocks, and np.block does '
+            'not allow implicit conversion from tuple to ndarray.'.format(
+                format_index(parent_index)
+            )
+        )
+    elif type(arrays) is list and len(arrays) > 0:
+        indexes = [_block_check_depths_match(arr, parent_index + [i])
+                   for i, arr in enumerate(arrays)]
+
+        first_index = indexes[0]
+        for i, index in enumerate(indexes):
+            if len(index) != len(first_index):
+                raise ValueError(
+                    "List depths are mismatched. First element was at depth "
+                    "{}, but there is an element at depth {} ({})".format(
+                        len(first_index),
+                        len(index),
+                        format_index(index)
+                    )
+                )
+        return first_index
+    elif type(arrays) is list and len(arrays) == 0:
+        # We've 'bottomed out' on an empty list
+        return parent_index + [None]
+    else:
+        # We've 'bottomed out' - arrays is either a scalar or an array
+        return parent_index
+
+
+def _block(arrays, max_depth):
+    def atleast_nd(a, ndim):
+        # Ensures `a` has at least `ndim` dimensions by prepending
+        # ones to `a.shape` as necessary
+        return array(a, ndmin=ndim, copy=False, subok=True)
+
+    def block_recursion(arrays, depth=0):
+        if type(arrays) is list:
+            if len(arrays) == 0:
+                raise ValueError('Lists cannot be empty')
+            arrs = [block_recursion(arr, depth+1) for arr in arrays]
+            arr_ndim = max(arr.ndim for arr in arrs)
+            arrs = [atleast_nd(a, arr_ndim) for a in arrs]
```

Comment:
Not a huge fan of how `atleast_nd` gets called repeatedly, rather than just once for each array.

Reply:
The simplest way to get around this would be to have _block_check_depths_match also calculate the max array ndim. I'll have a go at drafting that change and see how it looks and whether it improves performance.